modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -331,6 +431,166 @@ public void onFailure(Exception e) {
+                                    + "Retrying on the next DLM run",
+                                targetIndex
+                            );
+                            // We need to unset the force merge marker in the cluster state so that DLM will try force merging again
+                            unsetForceMergeMarker(targetIndex);
+                        } else {
+                            logger.error(
+                                () -> Strings.format(
+                                    "DLM encountered an error trying to force merge index [%s], and will not retry",
+                                    targetIndex
+                                ),
+                                e
+                            );
+                        }
+                        listener.onFailure(e);
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                listener.onFailure(e);
+            }
+        };
+
+        setForceMergeMarkerBegin(targetIndex, forceMergeRunningListener);
+    }
+
+    /*
+     * This method sets the value of the custom index metadata field "force_merge" within the field "dlm" to "complete". The method returns
+     * immediately, but the update happens asynchronously. Success or failure is logged.
+     */
+    private void setForceMergeMarkerComplete(String targetIndex) {
+        // Nothing actually waits for the result of this unsetting, so the listener is just used for logging:
+        ActionListener<Void> listener = new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.trace("Successfully set the force merge marker to complete on {}", targetIndex);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(() -> Strings.format("Could not set the force merge marker to complete on {}", targetIndex), e);
+            }
+        };
+        setForceMergeMarker(targetIndex, listener, FORCE_MERGE_COMPLETE_MARKER);
+    }
+
+    /*
+     * This method sets the value of the custom index metadata field "force_merge" within the field "dlm" to "begin". The method returns
+     * immediately, but the update happens asynchronously and listener is notified on success or failure.
+     */
+    private void setForceMergeMarkerBegin(String targetIndex, ActionListener<Void> listener) {
+        setForceMergeMarker(targetIndex, listener, FORCE_MERGE_BEGIN_MARKER);
+    }
+
+    /*
+     * This method sets the value of the custom index metadata field "force_merge" within the field "dlm" to value. The method returns
+     * immediately, but the update happens asynchronously and listener is notified on success or failure.
+     */
+    private void setForceMergeMarker(String targetIndex, ActionListener<Void> listener, String value) {
```

Comment:
I think we would benefit from having the value for these markers be timestamps (we'd be able to see _when_ forcemerge completed for an index). 
What do you think?

Reply:
That's a good idea.